### PR TITLE
v3.0.0

### PR DESCRIPTION
**Pull request by koji-1009**
- target & compile sdk 30
- refactor public and internal classes
- refactor tests

### CODE REVIEW

File: sdk-example-kt/src/main/java/jp/studyplus/android/sdk_example_kt/MainActivity.kt [hunk 1]
```diff
@@ -52,19 +57,15 @@ class MainActivity : AppCompatActivity() {
...
-                    })
+                instance.postRecord(record, object : PostCallback {
+                    override fun onSuccess() {
+                        Toast.makeText(context, "Post Study Record!!", Toast.LENGTH_LONG).show()
+                    }
+
+                    override fun onFailure(e: StudyplusError) {
+                        Toast.makeText(context, "error!!", Toast.LENGTH_LONG).show()
+                    }
+                })
```
Comment (nacatl) [inline sdk-example-kt/src/main/java/jp/studyplus/android/sdk_example_kt/MainActivity.kt:68]: Separation between an `onSuccess` and an `onFailure` callbacks 👍

File: studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/Studyplus.kt [hunk 2]
```diff
@@ -5,105 +5,89 @@ import android.content.Context
...
+import jp.studyplus.android.sdk.internal.auth.Preference
+import jp.studyplus.android.sdk.internal.auth.Transit
 import jp.studyplus.android.sdk.record.StudyRecord
-import java.io.IOException
 
-class Studyplus private constructor() {
+class Studyplus(
+    context: Context,
+    private val consumerKey: String,
+    private val consumerSecret: String,
```
Comment (nacatl) [inline studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/Studyplus.kt:16]: 👍

File: studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/StudyplusError.kt [hunk 3]
```diff
@@ -0,0 +1,18 @@
+package jp.studyplus.android.sdk
+
+sealed class StudyplusError {
+    class IOException(val throwable: Throwable) : StudyplusError()
+    object BadRequest : StudyplusError()
+    object LoginRequired : StudyplusError()
+    object ServerError : StudyplusError()
+    object Unknown : StudyplusError()
```
Comment (nacatl) [inline studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/StudyplusError.kt:8]: 👍👍👍

File: studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/internal/auth/Preference.kt [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+                "StudyplusSDK",
+                "The data is invalid. If code is correct, please contact Studyplus Dev team.",
+            )
+            return
+        }
+
+        if (code != RESULT_CODE_AUTHENTICATED) {
+            // not authenticated
+            return
+        }
```
Comment (nacatl) [inline studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/internal/auth/Preference.kt:33]: I think the SDK should return a log also here.

e.g. 
```kotlin
Log.e(
    "StudyplusSDK",
    "Your app is not authenticated.",
)
```

File: studyplus-android-sdk/src/test/java/jp/studyplus/android/sdk/StudyRecordUnitTest.kt [hunk 5]
```diff
@@ -11,22 +11,24 @@ import org.robolectric.RobolectricTestRunner
...
 import java.time.ZoneOffset
-import java.time.format.DateTimeFormatter
 
 @RunWith(RobolectricTestRunner::class)
 @Config(sdk = [Build.VERSION_CODES.P])
 class StudyRecordUnitTest {
 
     private val date = OffsetDateTime.of(2019, 6, 1, 1, 2, 3, 0, ZoneOffset.UTC)
-    private val dateStr = date.format(DateTimeFormatter.ISO_OFFSET_DATE_TIME)
+    private val dateStr = "2019-06-01T01:02:03Z"
```
Comment (nacatl) [inline studyplus-android-sdk/src/test/java/jp/studyplus/android/sdk/StudyRecordUnitTest.kt:20]: 👍

**Review by nacatl — COMMENTED** [review]
I commented a little

File: studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/internal/auth/Preference.kt (hunk 4 above)
Comment (koji-1009) [reply studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/internal/auth/Preference.kt]: I added 6a61fd6 !

**Review by nacatl — APPROVED** [review]
LGTM！

File: studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/internal/auth/Preference.kt (hunk 4 above)
Comment (nacatl) [reply studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/internal/auth/Preference.kt]: Thank you for your rapid response!!

File: studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/StudyplusError.kt (hunk 3 above)
Comment (JASON13F) [reply studyplus-android-sdk/src/main/java/jp/studyplus/android/sdk/StudyplusError.kt:8]: 👍

**Review by JASON13F — APPROVED** [review]
LGTM

**Pull request merged**